### ISSUE TITLE
Accessibility issues

### ISSUE THREAD

**Issue by arnaudlevy** [CloudCannon/pagefind#727]
Hi everyone!
Thanks a lot for the wonderful job with Pagefind :)
We use it as the default search engine in [Osuny](https://osuny.org), as you can see in the [example website](https://example.osuny.org/).
We benefited from an accessibility audit by [Temesis](https://www.temesis.com), funded by [Région Nouvelle-Aquitaine](https://www.nouvelle-aquitaine.fr/), in France. 
The audit is against [RGAA](https://accessibilite.numerique.gouv.fr/), which is french equivalent for WCAG.
It would be wonderful if we could make Pagefind fully RGAA compliant by default!

## Issues 

There are 3 issues, 1 blocking, 2 major.
I translate the issues and the fixes suggested by Temesis here.

- [ ] Blocking : No visible tag on search field [11.1](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#11.1)
- [ ] Major : Presence of an ordered list with no content [9.3](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#9.3)
- [ ] Major : Absence of restitution of the number of results [7.5](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#7.5)

## No visible tag on search field
> Either add a title attribute with a proper value on the field (same as placeholder), or us the floating label technique.

## Presence of an ordered list with no content
> Remove the empty `ol`

## Absence of restitution of the number of results
> Implement the  element `<div class=”resultat_saisie” aria-live=”polite”></div>` on page load, with no content inside

> Integrate the content when the results appear

On page load:
```html
<div class="resultat_saisie" aria-live="polite"></div>
```
When results come:
```html
<div class="resultat_saisie" aria-live="polite"><p>10 results for example</p></div>
```
When no results:
```html
<div class="resultat_saisie" aria-live="polite"><p>No results for dfgdfgdfg</p></div>
```

https://github.com/osunyorg/theme/issues/678

**Comment by bglw** [comment]
Hi @arnaudlevy 👋 

Thanks so much for sharing this! I'll make sure to action these points very soon.

**Comment by bglw** [comment]
👋 @rdela 

I always like PRs! Apologies for not getting to these fixes sooner — I need to dedicate some time to this. Let me know if you need any help with a PR for part of it, but also no pressure to contribute 🙂

**Comment by rdela** [comment]
Thanks @bglw! looks straightforward enough for me to get something opened. Especially since on further inspection the 1 blocking issue Arnaud identified originally is the same one I metioned :)

**Comment by rdela** [comment]
@arnaudlevy 

You’ll be happy to know I have landed an [initial commit](https://github.com/CloudCannon/pagefind/pull/798/commits/f2c164de7bbfe4b68c448cd93693a1c8a5a3565d) on #798 which I think would already solve the first Blocking issue you identify, “No visible tag on search field [11.1](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#11.1).” We have “add[ed] a title attribute with a proper value on the field (same as placeholder),” as you suggested.

I still need to document my build battles there, but in the meantime, wanted to ask here: Are you able to help me understand how to remedy the other issues better please? 

I read through the discussion at [Accessibilité de la recherche Pagefind (Pagefind search accessibility) - osunyorg/theme #678](https://github.com/osunyorg/theme/issues/678), and with the help of DeepL and Google Translate, expanded my understanding some already.

# RGAA 7.5

> Absence of restitution of the number of results [7.5](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#7.5)

>> In each web page, are [status messages ](https://accessibilite.numerique.gouv.fr/methode/glossaire/#message-de-statut) correctly returned by assistivery technologies?
>> Dans chaque page web, les [messages de statut](https://accessibilite.numerique.gouv.fr/methode/glossaire/#message-de-statut) sont-ils correctement restitués par les technologies d’assistance?

Earlier in this thread, you wrote (or pasted from the audit?): 

> Implement the element 
> `<div class=”resultat_saisie” aria-live=”polite”></div>` 
> on page load, with no content inside
> Integrate the content when the results appear

Maybe something is getting lost in translation, but I tend to agree more with what [you wrote on osunyorg/theme #678](https://github.com/osunyorg/theme/issues/678#issuecomment-2415888785), which I noticed was written later on that day: 

> Sur le 3e point, il me semble que nous contrevenons aux points 7.5.1 ou 7.5.3 parce qu'il n'y a pas d'attribut role.
> En revanche, je ne vois rien qui oblige à avoir un div vide.
> 
> Par ailleurs, l'aria-live polite vaut role=log, ce qui est pertinent pour un message de progression.
> Mais c'est faux, il s'agit du résultat d’une action, donc il faut :
> 
> soit role="status"
> soit aria-live="polite" et un attribut WAI-ARIA aria-atomic="true"
> Je propose d'ajouter role="status" sur la balise existante :
> 
> `<p class="pagefind-ui__message svelte-e9gkc3" role="status">52 résultats pour test</p>`

Or in Googlized English…

> On the third point, it seems to me that we're violating points 7.5.1 or 7.5.3 because there's no role attribute.
> 
> However, I don't see anything that requires an empty div.
> 
> Furthermore, the aria-live polite value is set to role=log, which is relevant for a progress message.
> But that's wrong; this is the result of an action, so we need:
> 
> either role="status"
> or aria-live="polite" and a WAI-ARIA attribute aria-atomic="true"
> I suggest adding role="status" to the existing tag:
> 
> `<p class="pagefind-ui__message svelte-e9gkc3" role="status">52 results for test</p>`

Therefore I propose adding `role="status"` to `<p class="pagefind-ui__message">` on [line 300 here in ui.svelte](https://github.com/CloudCannon/pagefind/blob/248f81d172316a642a83527fa92180abbb7f9c49/pagefind_ui/default/svelte/ui.svelte#L300) as you suggest there. Or did you decide `aria-live="polite"` + `aria-atomic="true"` was the way after all? If so, do we indeed need both? Please clarify what you think the best code change is here.

# RGAA 9.3

> Presence of an ordered list with no content [9.3](https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#9.3)

>> Dans chaque page web, chaque liste est-elle correctement structurée?
>> In each web page, is each list correctly structured?

Suggesting we are in violation of: 

>> La liste utilise les balises HTML `<ol>` et` <li>` ;
>> The list uses HTML tags `<ol>` and `<li>` ;

Again, I found [your osunyorg/theme #678 comment](https://github.com/osunyorg/theme/issues/678#issuecomment-2415865115) helpful, which was also written after the initial comment here:

> En 9.3.2, le RGAA précise qu'il faut utiliser des ol et des li.
> Il est tout à fait correct d'utiliser un ol vide pour indiquer une liste vide (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol).
> 
> Capture d’écran 2024-10-16 à 08 43 35
> Je pense que la préconisation est erronée, parce que:
> 
> - une liste vide est plus juste sémantiquement qu'une absence de liste
> - une liste vide est correcte du point de vue HTML
> - la méthodologie du test 9.3.2 ne traite pas le cas zero-li, ni dans un sens ni dans l'autre
> 
> Il me semble qu'il s'agit là d'une interprétation fausse de la phrase La liste utilise les balises HTML `<ol>` et `<li>`.
> Je propose de se conformer à la spec HTML.

Or in Googlized English…

> In 9.3.2, the RGAA specifies that ol and li must be used.
> It is perfectly correct to use an empty ol to indicate an empty list (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol).
> 
> Screenshot 2024-10-16 at 08:43:35
> I think the recommendation is wrong because:
> 
> - an empty list is more semantically accurate than the absence of a list
> - an empty list is HTML-correct
> - the methodology of test 9.3.2 does not address the zero-li case, either way
> 
> It seems to me that this is a misinterpretation of the sentence The list uses the HTML tags `<ol>` and `<li>`.
> I propose to comply with the HTML specification.

So, what is the verdict here? Leave the empty `<ol>` alone? Or wrap that section of the svelte markup with `{#if searchResult.results.length}` (or maybe more verbosely `{#if searchResult.results.length > 0}` for clarity), so that the `<ol>` only appears with results?

**Comment by arnaudlevy** [comment]
@rdela thanks!
Let me add @alexisben to the discussion, he knows in depth 😃

**Comment by rdela** [comment]
Great, thanks Arnaud and hello Alexis! Look forward to your thoughts!

For people following along who may be curious, I have documented my build battles at #798 now, starting with [code changes](https://github.com/CloudCannon/pagefind/pull/798#issuecomment-2762554159), and subsequently moving on to [browser testing](https://github.com/CloudCannon/pagefind/pull/798#issuecomment-2762909249), including how to use Headless Chromium aka `chrome-headless-shell` with toolproof if you do not want to install Google Chrome on your Mac.

**Comment by alexisben** [comment]
Hi @rdela! Sorry for the delay!

Here are the decisions we have taken that have been validated as compliant by the audit : 

### RGAA 7.5

We choose to had `aria-live="polite"` + `aria-atomic="true"` with a inner `<p>` tag added with javascript. 

```
  <div aria-live="polite" aria-atomic="true" class="sr-only pagefind-ui__accessible-message">
    <p>64 résultats pour test</p>
  </div>
```

With this, screen reader will read the informations without changing the tabulation position.


### RGAA  9.3

For this point, after a conversation with Temesis, keep an empty `<ol>` isn't a conformity issue but an amelioration. To handle this case, we hide the list with `css` when it's empty : 

```
#search .pagefind-ui__results:empty {
  display: none;
}
```

We thought if an `aria-polite` declare to screen reader there's no results to the search, it doesn't seems confusing to have an empty list. 


The html looks like (I removed some attributes or tags to make the code more readable) : 

```
<div id="search" aria-hidden="false" aria-modal="true" role="dialog" aria-label="Recherche">
  <button aria-label="Fermer la recherche"></button>
  <p role="heading" aria-level="1" class="sr-only">Rechercher</p>

  <div class="pagefind-ui pagefind-ui--reset">
    <form class="pagefind-ui__form" role="search" aria-label="Recherche sur ce site">

      <input class="pagefind-ui__search-input" type="text"
        placeholder="Rechercher" autocapitalize="none" enterkeyhint="search" title="Rechercher">
      <button class="pagefind-ui__search-clear">Nettoyer</button>

      <div class="pagefind-ui__drawer">
        <div class="pagefind-ui__results-area">
          <p class="pagefind-ui__message" aria-hidden="true">64 résultats pour test</p>
          <ol class="pagefind-ui__results">
            [...]
          </ol>
          <button type="button" class="pagefind-ui__button">Charger plus de résultats</button>
        </div>
      </div>
    </form>
  </div>
  <div aria-live="polite" aria-atomic="true" class="sr-only pagefind-ui__accessible-message">
    <p>64 résultats pour test</p>
  </div>
</div>
```

**Comment by bglw** [comment]
Hey all — the latest [Pagefind Beta](https://github.com/Pagefind/pagefind/releases/tag/v1.5.0-beta.1) is set to replace the Default UI, and by far the most work went into accessibility-hardening it. In saying that, I'm sure there are things that can be improved! It's at least leaps and bounds better in this facet, and others.

I'll drop another note here once it has stablized, and likely close this issue out as the Default UI becomes deprecated. (And welcoming any issues to be opened, and addressed substantially quicker, for the new Component UI!)

**Comment by steveoh** [comment]
I am so excited for this! Great work

**Comment by alexisben** [comment]
@bglw many thanks and congrats! We plan to test the beta into our [Osuny Theme](https://github.com/osunyorg/theme)

**Comment by bglw** [comment]
Component UI is out in [v1.5.0](https://github.com/Pagefind/pagefind/releases/tag/v1.5.0) — closing this one here for now but feel free to open fresh, more urgent issues for any Component UI fixes to be made!
